### ISSUE TITLE
Add wall/ground awarness to Projectile Prediction

### ISSUE THREAD

**Issue by SylveonBottle** [nullworks/cathook#1177]
Pre-"optimized" NullCore (2017 I think?) was probably the first and only ever chear that could take in account of the target moving up or down slopes "along his path", or even take account for a player more than likely hitting a wall if they were rocket jumping.

Yes, cool, nearly all cheats' projectile aimbot can hit targets flung at an exact unmoving path with no obstacles on tr_walkway. Even better, some cheats can take account for the target _landing_ on solid, flat ground, and shoot where they'll land.

But, here's the thing. Rockets travel slowly, and people have physics. The average target isn't just going to stop moving the moment they land- they'll probably keep moving in whatever direction they're trying to move in. A person walking forwards towards a slope isn't going to go _through_ the slope, they're going to scale up it. And an enemy walking towards the wall isn't going to go through the wall either, so stop targeting as if they have noclip and walls don't exist. Looking at you, lbox.

If medic continues on his current direction, he'll _hit the ground_ and continue forward. 
![image](https://user-images.githubusercontent.com/17802843/100302491-0e681480-2f60-11eb-8df5-7608f5efe0f9.png)

If the heavy was another class that wasn't slow and reved up, and the user wasn't using the direct hit, and was also a little further away (don't judge me it's incredibly hard to find pre-2017 NCC footage with prediction lines enabled).. We should definitely account for the fact that the target is about to walk off a ledge. 
![image](https://user-images.githubusercontent.com/17802843/100303010-2d1adb00-2f61-11eb-96b7-1716df7fa8fe.png)

Self-prediction with horrible quality because, hey guess what, it's incredibly hard to find footage for this.
But look at how NCC used to take account for **walls.** 
![image](https://user-images.githubusercontent.com/17802843/100303251-b5997b80-2f61-11eb-8765-298999117fd4.png)

I was literally in a game of lbox a few days ago and I remember this engineer was behind a wall real far away from me. There was no way for him to get to where I was except for a window he could climb up. He was just minding his own business, walking around building his little nest up.. _And lmaobox locked on him through the fucking wall._ It thought he could go _through the wall_. Maybe it was taking account for the window, but there's no way that dude could go through walls. And let lmaobox doesn't take account of walls, so..

Basically, what I'm wanting is some "movement prediction" thing that allows projectile prediction to take account of the target colliding with walls and the target changing Y levels (from either going up or down slopes, or falling off ledges). Yes, it's super cool that you can predict strafing now. But if they strafe into a wall, it's going to try and shoot inside of the wall, ain't it? 

Bonus points if the server sends us what WASD / direction the user is walking while mid-air. If we knock a pyro into the air cleanly, and he starts holding backwards while falling, chances are he's going to go backwards when he lands.

**Comment by SylveonBottle** [comment]
I'd like to add a perfectly good example as to how accounting for walls mid-air is important. 

https://youtu.be/D2GtFiw9HyA?t=87

Look at this fucking scout. Look at this shitty proj aimbot with this fucking scout.
First rocket hits the scout in the chest, knocking him backwards. Okay, that's fine- personally I would've made the proj aimbot aim at the feet so we can drag his airtime out for as long as possible, but whatever. I guess shooting at their chest deals a little more damage with direct hits.
Second rocket.. The SECOND ROCKET. THIS FUCKER IS _SHOCKED_ AT THE IDEA THAT THE SCOUT HITS THE WALL BEFORE THE ROCKET CAN REACH IT.
The scout. Hits a fucking wall. And the aimbot misses. Because the aimbot didn't predict for the wall, and thought scouts could go through bounding boxes for some reason.
This happens _all the time_ in casual gameplay. How many times does a person "hit a wall" while they're mid-air? They do it to slow themselves down, they do it to rocketjump faster, they do it because you hit a nice rocket or pill on them and flung them _anywhere except directly up_ and they more than likely hit a wall. 
This fucking projectile prediction... didn't account for the wall. So it _missed_. 
Please do not tell me that cathook projectile aimbot has this flaw as well. And if it does, please take this issue into consideration for improving proj aimbot.

**Comment by TotallyNotElite** [comment]
OK so half this issue is invalid because cathook's aimbot already takes the floor into consideration.

Regarding the walls thing: Predicting anything walls related is going to be rather expensive and it doesn't occur _that_ often. Maybe a simple vischeck from the start to the end location could be considered?

**Comment by senator715** [comment]
> 
> 
> OK so half this issue is invalid because cathook's aimbot already takes the floor into consideration. Idk why you even talked about that, because cathook literally does that already.
> 
> Regarding the walls thing: Predicting anything walls related is going to be rather expensive and it doesn't occur _that_ often. Maybe a simple vischeck from the start to the end location could be considered?

Sounds to me like you are just avoiding 3 lines of code, its not resourceful at all, you just don't know how the collision logic works, go read the movement code and you will see it right in front of you.

Also "it doesn't occur that often" is completely false.

**Comment by TotallyNotElite** [comment]
@senator715 thanks for your insightful comment. I have actually already written cheap code that does correctly take walls into consideration, I'm just not happy with it's quality and have therefore decided to push it back to a later date. I appreciate your interest in this issue, but would like to inform you that it is not your choice what I spend my free time on.

Sincerely, TNE.
